rust/lancedb/src/database/listing/manifest.rs

```diff
@@ -0,0 +1,3991 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The LanceDB Authors
+
+//! Manifest table management for v2 listing databases
+
+use crate::database::listing::{
+    apply_pagination, apply_pagination_with_key, ListingDatabase, NewTableConfig,
+    LANCE_FILE_EXTENSION,
+};
+use crate::database::{
+    CloneTableRequest, CreateNamespaceRequest, CreateTableMode, Database, DropNamespaceRequest,
+    ListNamespacesRequest, ReadConsistency,
+};
+use crate::error::{Error, Result};
+use crate::table::dataset::DatasetConsistencyWrapper;
+use crate::table::{BaseTable, NativeTable};
+use crate::utils::{validate_namespace, validate_table_name};
+use arrow_array::{Array, ListArray, RecordBatch, RecordBatchIterator, StringArray};
+use arrow_schema::{DataType, Field, Schema};
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use lance::dataset::optimize::{compact_files, CompactionOptions};
+use lance::dataset::refs::Ref;
+use lance::dataset::{builder::DatasetBuilder, InsertBuilder, ReadParams, WriteMode, WriteParams};
+use lance::session::Session;
+use lance::Dataset;
+use lance_datafusion::utils::StreamingWriteSource;
+use lance_index::optimize::OptimizeOptions;
+use lance_index::scalar::ScalarIndexParams;
+use lance_index::{DatasetIndexExt, IndexType};
+use lance_io::object_store::{ObjectStore, ObjectStoreParams, WrappingObjectStore};
+use lance_table::io::commit::commit_handler_from_url;
+use rand::Rng;
+use std::cmp::PartialEq;
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+use std::hash::{DefaultHasher, Hash, Hasher};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Configuration for the manifest listing database
+#[derive(Debug, Clone)]
+pub struct ManifestListingDatabaseConfig {
+    /// Read consistency interval for the manifest table
+    pub read_consistency_interval: Option<Duration>,
+    /// Whether to enable inline optimization (compaction and indexing) on the manifest table
+    pub inline_optimization_enabled: bool,
+    /// Whether directory listing is enabled in the parent database
+    /// If true, new tables in root namespace will keep the <table_name>.lance naming.
+    /// If false, they can use hash-based names without .lance extension.
+    pub parent_dir_listing_enabled: bool,
+}
+
+impl Default for ManifestListingDatabaseConfig {
+    fn default() -> Self {
+        Self {
+            read_consistency_interval: None,
+            inline_optimization_enabled: true,
+            parent_dir_listing_enabled: true, // Default to true for backward compatibility
+        }
+    }
+}
+
+/// Delimiter used in object IDs to separate namespace components and table names
+const DELIMITER: &str = "$";
```

Comment:
~~Is this a reserved character that we shouldn't use in table names now?~~

Nevermind, I see it was already a reserved character for tables names and namespaces!